### ISSUE TITLE
Reducing mining speed with origins:modify_break_speed power doesnt let you break instabreak blocks 

### ISSUE THREAD

**Issue by ChrysanthCow** [apace100/apoli#170]
See: EdwinMindcraft/origins-architectury#379

That report is on 1.19.2, but, it is replicable over here.

**Comment by eggohito** [comment]
I'm not sure on what would be the best way of fixing this issue. On one hand, it can be fixed by supplying a base hardness value if the hardness value of the block is 0 (which is the case for blocks that can be insta-mined), but on the other hand, the user can account for such blocks using the `hardness` block condition type.

If the latter is to be done, the user themselves can supply a base hardness value for insta-mineable blocks in the modifier for the `modify_break_speed` power type, and if the former is to be done, the supplied value will either be a constant or a value controlled by an additional field (similar to how the `explode` action types now have an `indestructible_resistance` field for controlling the blast resistance value of blocks that are considered indestructible)

**Comment by ChrysanthCow** [comment]
That's fair, whatever you come up with, I think it'll be good.

**Comment by 0vergrown** [comment]
Workaround: I ran into the same issue with my Origin and the workaround I came up with was adding all the instabreak blocks into a tag:

```
{
    "name": "Slow Miner",
    "description": "Their small stature makes mining slower for Inchlings.",
    "type": "origins:modify_break_speed",
    "block_condition": {
        "type": "origins:in_tag",
        "tag": "improved_inchling:breakable",
        "inverted": true
    },
    "modifier": {
        "operation": "multiply_base",
        "value": -0.6
    }
}
```

```
{
    "values": [
      "minecraft:beetroots",
      "minecraft:carrots",
      "minecraft:potatoes",
      "minecraft:wheat",
      "minecraft:melon_stem",
      "minecraft:pumpkin_stem",
      "minecraft:torchflower_crop",
      "minecraft:pitcher_crop",
      "minecraft:tube_coral",
      "minecraft:brain_coral",
      "minecraft:bubble_coral",
      "minecraft:fire_coral",
      "minecraft:horn_coral",
      "#minecraft:coral_plants",
      "minecraft:tube_coral_fan",
      "minecraft:brain_coral_fan",
      "minecraft:bubble_coral_fan",
      "minecraft:fire_coral_fan",
      "minecraft:horn_coral_fan",
      "#minecraft:small_flowers",
      "#minecraft:tall_flowers",
      "minecraft:flowering_azalea_leaves",
      "minecraft:flowering_azalea",
      "minecraft:mangrove_propagule",
      "minecraft:cherry_leaves",
      "minecraft:pink_petals",
      "minecraft:flower_pot",
      "minecraft:potted_poppy",
      "minecraft:potted_blue_orchid",
      "minecraft:potted_allium",
      "minecraft:potted_azure_bluet",
      "minecraft:potted_red_tulip",
      "minecraft:potted_orange_tulip",
      "minecraft:potted_white_tulip",
      "minecraft:potted_pink_tulip",
      "minecraft:potted_oxeye_daisy",
      "minecraft:potted_dandelion",
      "minecraft:potted_oak_sapling",
      "minecraft:potted_spruce_sapling",
      "minecraft:potted_birch_sapling",
      "minecraft:potted_jungle_sapling",
      "minecraft:potted_acacia_sapling",
      "minecraft:potted_dark_oak_sapling",
      "minecraft:potted_red_mushroom",
      "minecraft:potted_brown_mushroom",
      "minecraft:potted_dead_bush",
      "minecraft:potted_fern",
      "minecraft:potted_cactus",
      "minecraft:potted_cornflower",
      "minecraft:potted_lily_of_the_valley",
      "minecraft:potted_wither_rose",
      "minecraft:potted_bamboo",
      "minecraft:potted_crimson_fungus",
      "minecraft:potted_warped_fungus",
      "minecraft:potted_crimson_roots",
      "minecraft:potted_warped_roots",
      "minecraft:potted_azalea_bush",
      "minecraft:potted_flowering_azalea_bush",
      "minecraft:potted_mangrove_propagule",
      "minecraft:potted_cherry_sapling",
      "minecraft:potted_torchflower",
      "minecraft:oak_sapling",
      "minecraft:spruce_sapling",
      "minecraft:birch_sapling",
      "minecraft:jungle_sapling",
      "minecraft:acacia_sapling",
      "minecraft:dark_oak_sapling",
      "minecraft:azalea",
      "minecraft:cherry_sapling",
      "minecraft:dandelion",
      "minecraft:poppy",
      "minecraft:blue_orchid",
      "minecraft:allium",
      "minecraft:azure_bluet",
      "minecraft:red_tulip",
      "minecraft:orange_tulip",
      "minecraft:white_tulip",
      "minecraft:pink_tulip",
      "minecraft:oxeye_daisy",
      "minecraft:cornflower",
      "minecraft:lily_of_the_valley",
      "minecraft:wither_rose",
      "minecraft:torchflower",
      "minecraft:sunflower",
      "minecraft:lilac",
      "minecraft:peony",
      "minecraft:rose_bush",
      "minecraft:pitcher_plant",
      "minecraft:tube_coral_wall_fan",
      "minecraft:brain_coral_wall_fan",
      "minecraft:bubble_coral_wall_fan",
      "minecraft:fire_coral_wall_fan",
      "minecraft:horn_coral_wall_fan",
      "minecraft:sugar_cane",
      "minecraft:attached_pumpkin_stem",
      "minecraft:attached_melon_stem",
      "minecraft:red_mushroom",
      "minecraft:brown_mushroom",
      "minecraft:warped_fungus",
      "minecraft:warped_roots",
      "minecraft:crimson_fungus",
      "minecraft:crimson_roots",
      "minecraft:grass",
      "minecraft:tall_grass",
      "minecraft:seagrass",
      "minecraft:tall_seagrass",
      "minecraft:torch",
      "minecraft:wall_torch",
      "minecraft:soul_torch",
      "minecraft:soul_wall_torch",
      "#minecraft:bee_growables",
      "minecraft:tripwire",
      "minecraft:hanging_roots"
    ]
  }
```

This solved the issue, so maybe do something similar? It's the only thing I've been able to come up with that works.

**Comment by ChrysanthCow** [comment]
The above should not be the permanent solution to this problem though. Requiring adding all instabreak blocks to a tag is pretty hefty in terms of workload, and not everybody would've thought of it.

**Comment by 0vergrown** [comment]
I am aware, I thought it could help maybe come up with an idea that could be a permanent solution. Hence its a `workaround`

**Comment by eggohito** [comment]
> Workaround: I ran into the same issue with my Origin and the workaround I came up with was adding all the instabreak blocks into a tag...

You don't necessarily need a block tag, since you can just use the [`hardness` block condition type](https://origins.readthedocs.io/en/latest/types/block_condition_types/hardness) to check if the block has a hardness value of `>` `0` as I mentioned in my previous comment (with the exception of being able to supply a replacement hardness value since the calculated block breaking delta value reaches (positive/negative) infinity or NaN)

This workaround eliminates the need of having to list blocks that can be insta-mined in a block tag manually, since blocks that can be insta-mined have a hardness value of `0`

```json5
{
    type: "apoli:modify_break_speed",
    block_condition: {
        type: "apoli:hardness",
        comparison: ">",
        compare_to: 0
    },
    modifiers: {
        operation: "multiply_total",
        value: -0.9
    }
}
```

<br>

**EDIT:** You can also use a modifier with the `max_base`/`max_total` operation and a value of `Infinity` (if you're using JSON5) or `9218868437227405312` (if you're using JSON). This essentially replaces the calculated block breaking delta value to positive infinity, when it was NaN previously

```json5
{
    type: "apoli:modify_break_speed",
    modifiers: [
        {
            operation: "multiply_total",
            value: -0.9
        },
        {
            operation: "max_total",
            value: Infinity
        }
    ]
}
```

**Comment by eggohito** [comment]
> That's fair, whatever you come up with, I think it'll be good.

I'm thinking of adding `hardness_modifier(s)` field(s), which would allow users to modify the block's *effective* hardness value when calculating the block breaking delta value, which is then bound to a minimum value of `-1.0` since that's the lowest hardness value used for unbreakable blocks

**Comment by eggohito** [comment]
> I'm thinking of adding `hardness_modifier(s)` field(s), which would allow users to modify the block's _effective_ hardness value when calculating the block breaking delta value, which is then bound to a minimum value of `-1.0` since that's the lowest hardness value used for unbreakable blocks

This should now be implemented as of 9a60fb41ca024eba9e5ecfa81e2b7d54d86680c0
